src/test/java/com/rackspace/ceres/app/web/QueryControllerTest.java

```diff
@@ -0,0 +1,111 @@
+public class QueryControllerTest {
+
+  @MockBean
+  QueryService queryService;
+
+  @Autowired
+  private WebTestClient webTestClient;
+
+  @Test
+  public void testQueryApi() {
+
+    Map<String, String> queryTags = new HashMap<>();
+    queryTags.put("os", "linux");
+    queryTags.put("deployment", "dev");
+    queryTags.put("host", "h-1");
+
+    Map<Instant, Double> values = new HashMap<>();
+    values.put(Instant.now(), 111.0);
+
+    List<QueryResult> queryResults = new ArrayList<>();
+    queryResults.add(new QueryResult().setMetricName("cpu-idle").setTags(queryTags).setTenant("t-1")
+        .setValues(values));
+
+    when(queryService.queryRaw(anyString(), anyString(), any(), any(), any()))
+        .thenReturn(Flux.just(queryResults.get(0)));
+
+    webTestClient.get()
+        .uri(uriBuilder -> uriBuilder.path("/api/query")
+            .queryParam("tenant", "t-1")
+            .queryParam("metricName", "cpu-idle")
+            .queryParam("tag", "os=linux")
+            .queryParam("start", "1d-ago")
+            .build())
+        .exchange().expectStatus().isOk()
+        .expectBodyList(QueryResult.class).isEqualTo(queryResults);
+  }
+
+  @Test
+  public void testQueryApiWithAggregator() {
+
+    Map<String, String> queryTags = new HashMap<>();
+    queryTags.put("os", "linux");
+    queryTags.put("deployment", "dev");
+    queryTags.put("host", "h-1");
+
+    Map<Instant, Double> values = new HashMap<>();
+    values.put(Instant.now(), 111.0);
+
+    List<QueryResult> queryResults = new ArrayList<>();
+    queryResults.add(new QueryResult().setMetricName("cpu-idle").setTags(queryTags).setTenant("t-1")
+        .setValues(values));
+
+    when(queryService.queryDownsampled(anyString(), anyString(), any(), any(), any(), any(), any()))
+        .thenReturn(Flux.just(queryResults.get(0)));
+
+    webTestClient.get()
+        .uri(uriBuilder -> uriBuilder.path("/api/query")
+            .queryParam("tenant", "t-1")
+            .queryParam("metricName", "cpu-idle")
+            .queryParam("tag", "os=linux")
+            .queryParam("start", "1d-ago")
+            .queryParam("end", "1605117336")
+            .queryParam("aggregator", "min")
+            .queryParam("granularity", "pt1m")
+            .build())
+        .exchange().expectStatus().isOk()
+        .expectBodyList(QueryResult.class).isEqualTo(queryResults);
+  }
+
+  @Test
+  public void testQueryApiWithAggregatorAndMissingGranularity() {
+
+    webTestClient.get()
+        .uri(uriBuilder -> uriBuilder.path("/api/query")
+            .queryParam("tenant", "t-1")
+            .queryParam("metricName", "cpu-idle")
+            .queryParam("tag", "os=linux")
+            .queryParam("start", "1d-ago")
+            .queryParam("end", "1605117336")
+            .queryParam("aggregator", "min")
+            .build())
+        .exchange().expectStatus().is5xxServerError()
+        .expectBody()
+        .jsonPath("$.status").isEqualTo(500)
```

Comment:
Hmm, I was expecting this to be a 400. Can you dig into that a little bit?

Reply:
I think we need to write custom exception handler for 400. As of now it defaults to 500

Reply:
Probably so. It's odd how some of the defaults like that aren't quite what you'd think. Can you write up a jira issue, put in the "refined" sprint near the top?

Reply:
Sure